fix(android): fix onprogress payload value for data other than HashMap

ypbnv opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27293

**Description:**
Request's length was not accounted for data of type different from HashMap, thus not reporting the
progress properly in the callback. Also there was a duplication in calculating a request's data size, so I removed the redundant variable. I believe all the cases are covered from the code that is left.
Working on a dedicated unit test for this.

Edit: Added the unit test.

**Test case:**
The JIRA ticket has a good code sample for testing.

ypbnv committed:
fix(android): fix onprogress payload value for data other than HashMap

request's length was not accounted for data of type different from HashMap, thus not reporting the
progress properly in the callback

build commented:
<!--
  2 failure:  :microscope: Ther..., Tests have failed...
  1 warning:  This PR has miles...
  4 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/ypbnv/titanium_mobile/blob/TIMOB-27293/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This PR has milestone set to 8.3.0, but the version defined in package.json is 8.2.0
Please either:
- Update the milestone on the PR
- Update the version in package.json
- Hold the PR to be merged later after a release and version bump on this branch</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:thumbsup: Hey!, You deleted more code than you added. That's awesome!</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-11168/1/artifact/dist/mobilesdk-8.2.0.v20190822082322-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 4 tests have failed
There are 4 tests failing and 472 skipped out of 4344 total tests.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.Titanium.Network.HTTPClient</td><td>callbackTestForPOSTMethod</td><td>60.014</td><td><pre>Error: timeout of 60000ms exceeded
at Titanium.&lt;anonymous&gt; (/ti-mocha.js:4326:18)</pre></td></tr>
<tr><td>android.emulator.Titanium.UI.TabGroup</td><td>#set_activeTab_in_creation_dictionary</td><td>9.916</td><td><pre>Error: timeout of 5000ms exceeded
at Titanium.&lt;anonymous&gt; (/ti-mocha.js:4326:18)</pre></td></tr>
<tr><td>ios.Titanium.UI.iOS</td><td>#createPreviewContext()</td><td>0.002</td><td><pre>Error: expected 0 to equal 300
fail@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/node_modules/should/cjs/should.js:356:23
file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti.ui.ios.previewcontext.test.js:26:48
file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:4376:41
file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:4763:17
file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:4840:23
next@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:4688:20
file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:4698:15
next@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:4636:30
file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:4665:13
timeslice@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti-mocha.js:5764:29</pre></td></tr>
<tr><td>ios.Titanium.UI.Window</td><td>.safeAreaPadding for window inside navigation window with extendSafeArea true</td><td>0.007</td><td><pre>Error: expected 34 to equal 0
fail@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/node_modules/should/cjs/should.js:275:23
value@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/node_modules/should/cjs/should.js:356:23
listener@file:///Users/build/Library/Developer/CoreSimulator/Devices/75DB2D1D-C851-40E3-84E2-69EDAAC907AE/data/Containers/Bundle/Application/7687729F-5BAB-4AE6-9FFA-D0D628C35C86/mocha.app/ti.ui.window.test.js:721:38</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against b3345a5ca808cf297a24468ef4d2e8dba0bd1d0d
</p>

garymathews committed:
Merge branch 'master' into TIMOB-27293

garymathews reviewed (approved):
CR: **PASS**

sgtcoolguy committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
test: missing unit test

ypbnv committed:
Merge branch 'master' into TIMOB-27293

jquick-axway commented on android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java:
```diff
@@ -1279,7 +1274,7 @@ public void run()
 							{
 								KrollDict data = new KrollDict();
-								double currentProgress = ((double) progress / totalLength);
+								double currentProgress = ((double) progress / contentLength);
```
@ypbnv, we should check if content length is zero. I can see this potentially happening with an HTTP "PUT" without a body.
```java
public void progress(int progress)
{
	if (contentLength <= 0) {
		return;
	}

	// The rest of the code goes here...
}
```

ypbnv replied:
Done.

Commit:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
fix(android): add guard for zero length request

Commit:
Merge branch 'master' into TIMOB-27293

ssekhri commented:
FR Passed.
Verified on:
Mac OS 10.14.6
Ti SDK: 8.3.0.v20190916012915
Appc CLI: 7.1.1
Node: 10.5.0
JDK: 10.0.2
Studio: 5.1.4.201909061933
Device: Nexus4 (v5.1.1), Pixel2 emulator (8.0.0)

sgtcoolguy committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
test(ios): trying to fix the unit test fir iOS

ypbnv committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
test(ios): try using ImageView for cross-platform test

jquick-axway commented:
@ypbnv, can you look into the unit test failure please? That's the only thing blocking this PR. Thanks!

ypbnv committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
test(ios): try to fix progress event unit test for iOS

garymathews commented:
Looks like this has broke `callbackTestForPOSTMethod` test

ypbnv commented:
After some more research - `callbackTestForPOSTMethod` should serve the same purpose of the unit test added by this PR. It should check the progress callback for sending data through HTTPRequest.

It fails because the test is successful when the progress event callback returns 1. This happened before, because it was calculated using the removed `totalLength` variable. It contained the length **only** of the data in the passed dictionary. In the case of the unit test that was a buffer with 10*1024 which has a 512 factor (this is how we track progress - with chunks of 512 bytes), so the last 512 bytes triggered the event with a progress 1. I think this would not be a reliable test because any content size different from X times of 512 bytes will never return progress value of 1. I am not really sure if this is the case for iOS, though.

ypbnv committed:
test(ios): put the send data in a dictionary

ypbnv committed:
test(ios): try to pass the iOS test for progress event

ypbnv committed:
test(ios): try to pass the iOS unit test

ypbnv committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
test(ios): try to pass a unit test in iOS

ypbnv committed:
Merge branch 'TIMOB-27293' of https://github.com/ypbnv/titanium_mobile into TIMOB-27293

ypbnv committed:
test(android): try to pass the unit test

jquick-axway commented:
Yes, the progress should be `1.0` (aka: `100%`) when the upload finishes successfully. That is a reasonable expectation. This problem needs to be solved.

jquick-axway commented:
@ypbnv, add the following code to the `ProgressOutputStream` class. This will solve the problem (I tested it out and it does the job).
```java
@Override
public void close() throws IOException
{
	super.close();

	if (!aborted && (transferred > 0)) {
		lastTransferred = transferred;
		listener.progress(transferred);
	}
}
```

Ideally, the progress handling should be redesigned, but I can live with this.

ypbnv committed:
Merge branch 'master' into TIMOB-27293

ypbnv committed:
fix(android): fix progress callback for request without multipart

remove redundant test file and ensure progress triggers 1 as a callback every time

ypbnv committed:
test: forgot to push the changed test

ypbnv commented:
@jquick-axway Added the snippet and also moved the closing of PrintWriter to happen for request that do not need multipart too.

jquick-axway reviewed (approved):
CR: Pass
